### DISCUSSION TITLE
Obtaining Contract Offer ID after Contract Definition Creation

### DISCUSSION THREAD

**Discussion by daiwai** [eclipse-edc/Connector#3409] [Q&A] [answered]
We are writing an EDC extension that is connected to an external catalogue which integrates data sets from various sources.

Our goal is to send new offers including the complete `ContractOfferId` to our external catalog as soon as a `Contract Definition` is created. At the moment we listen to the `ContractDefinitionCreated` event but as far as we could see the actual `Offers` are only created on the fly once the Connector receives a catalog request.

What would be the best way to handle our use case? Are there any extension points we could use to access the `Offer` after a `Contract Definition` has been created?

**Comment by jimmarino** [comment]
Offers are intended to be ephemeral and are scoped to the client.  Why would these need to be persisted on the provider?

**Comment by daiwai** [comment]
@jimmarino, thank you for your clarifying reply.

> Why would these need to be persisted on the provider?

To be clear, our initial thought wasn't to persist the offers on the provider but on a separate dedicated catalog, that would link consumers to the respective provider of an asset.

So if I understand correctly, we should rather only store the `Contract Definition Id` and `Asset Id` in the catalog and the client would then have to query the provider catalog with that data to generate an offer and initiate the contract negotiation?

Can you explain why you decided to make the offers ephemeral, or maybe point me to some resource which gives more background information on how to best realize a use case like ours?

**Reply by jimmarino** [reply, marked as answer]
Please look at [the Federated Catalog](https://github.com/eclipse-edc/FederatedCatalog), the FCC, and FCN. Also, I would read the [DSP protocol specifications](https://github.com/International-Data-Spaces-Association/ids-specification) and review the EDC documentation and collateral (which could be improved). For example:

https://github.com/eclipse-edc/Collateral/tree/main/Latest%20Presentations

**Reply by daiwai** [reply]
We are not using the Eclipse Federated Catalog but another one. Could you elaborate on why the offers are scoped to the client? Are different offers generated depending on which client queries the catalog?

**Reply by jimmarino** [reply]
Please look at how contract definitions work in EDC, in particular, access policy and the `assetsSelector`. If you are trying to cache offers in a third-party "broker" that is not advised as offers may be restricted by business rules such as, "only my gold customers should have access to this class of assets."

We're happy to help answer questions, but please carefully review the DSP specifications, EDC codebase, and try and raise specific issues.
